oresat_configs/scripts/pdo.py

```diff
@@ -0,0 +1,148 @@
+        help="oresat mission, defaults to %(default)s",
+    )
+    parser.add_argument("card", help="card name")
+    parser.add_argument(
+        "--list", action="store_true", help="list PDOs expected for the particular card"
+    )
+    parser.add_argument(
+        "--bus", default="vcan0", help="CAN bus to listen on, defaults to %(default)s"
+    )
+    return parser
+
+
+def register_subparser(subparsers):
+    """Registers an ArgumentParser as a subcommand of another parser.
+
+    Intended to be called by __main__.py for each script. Given the output of add_subparsers(),
+    (which I think is a subparser group, but is technically unspecified) this function should
+    create its own ArgumentParser via add_parser(). It must also set_default() the func argument
+    to designate the entry point into this script.
+    See https://docs.python.org/3/library/argparse.html#sub-commands, especially the end of that
+    section, for more.
+    """
+    parser = build_parser(subparsers.add_parser("pdo", help=PDO))
+    parser.set_defaults(func=pdo_main)
+
+
+typenames = {
+    canopen.objectdictionary.BOOLEAN: "bool",
+    canopen.objectdictionary.INTEGER8: "i8",
+    canopen.objectdictionary.INTEGER16: "i16",
+    canopen.objectdictionary.INTEGER32: "i32",
+    canopen.objectdictionary.UNSIGNED8: "u8",
+    canopen.objectdictionary.UNSIGNED16: "u16",
+    canopen.objectdictionary.UNSIGNED32: "u32",
+    canopen.objectdictionary.REAL32: "f32",
+    canopen.objectdictionary.VISIBLE_STRING: "str",
+    canopen.objectdictionary.OCTET_STRING: "bytes",
+    canopen.objectdictionary.UNICODE_STRING: "ustr",
+    canopen.objectdictionary.DOMAIN: "domain",
+    # canopen.objectdictionary.INTEGER24: 'i24',
+    canopen.objectdictionary.REAL64: "f64",
+    canopen.objectdictionary.INTEGER64: "i64",
+    # canopen.objectdictionary.UNSIGNED24: 'u24',
+    canopen.objectdictionary.UNSIGNED64: "u64",
+}
+
+
+def transmission_type(t: int) -> str:
+    """Retreives a name for a TPDO Transmission type
+
+    Subindex 2 of a PDO communication parameter record. See CiA-301 table 72.
+    """
+    if 0 <= t < 0xF0:
+        return f"sync ({t})"
+    if 0xF0 < t < 0xFB:
+        return "reserved"
+    if t == 0xFC:
+        return "RTR (sync)"
+    if t == 0xFD:
+        return "RTR (event)"
+    if t == 0xFE:
+        return "event (mfg)"
+    if t == 0xFF:
+        return "event (profile)"
+    raise ValueError(f"Invalid transmission type 0x{t:X}")
+
+
+def print_map(m: canopen.pdo.base.Map):
+    """Prints out a received PDO.
+
+    Which from this library means a PDO Mapping
+    """
+    data = []
+    for v in m:
+        signed = v.od.data_type in canopen.objectdictionary.SIGNED_TYPES
+        value = int.from_bytes(v.get_data(), byteorder="little", signed=signed)
+        data.append(f"{v.name}: {value}")
+    print(f'{m.cob_id:03X} {m.name} {" ".join(data)}')
+
+
+def listen(bus: str, node_id: int, od: canopen.ObjectDictionary):
```

Comment:
Maybe I'm not familiar with this enough, but is there something this is doing that we're not already achieving with SavvyCAN?

I was hoping to get away from any freewheeled CAN parsers across OreSat projects, since keeping a parser up-to-date is a lot of work.

This also seems out of scope from what configs was intended to be: a simple static config generator/manager.

Reply:
It's not a CAN parser, at least not directly. It goes through CANopen, specifically avoiding raw CAN messages. It assembles itself directly from the Object Dictionary generically, so there's not much to keep up-to-date. When the yaml configs get updated it gets updated automatically. As far as I can tell SavvyCAN isn't CANopen aware, the user needs to add their own definitions on it. 

My plan is to enhance this to be able to send as well, to a given card's RPDOs, and eventually add in RTR PDOs as well. Unless I missed something, SavvyCAN is capture only? 

As far as scope, this is the dual to `transfer-sdo.py`, which is in this project. Both the PDO and SDO scripts could be broken out into a separate repo, but that would be a small repo and the Oresat project already suffers from too many of them.